cranelift/module/src/module.rs

```diff
@@ -229,6 +229,22 @@ pub enum ModuleError {
     /// Wraps a `cranelift-codegen` error
     Compilation(CodegenError),
 
+    /// Memory allocation failure from a backend
+    Allocation {
+        /// Tell where the allocation came from
+        message: &'static str,
+        /// Io error the allocation failed with
+        err: std::io::Error,
+    },
+
+    /// Syscall failure from a backend
+    Syscall {
+        /// Tell what the syscall was for
+        message: &'static str,
+        /// Io error the syscall failed with
+        err: std::io::Error,
+    },
```

Comment:
I think this one should be placed behind the `Backend` variant.

Reply:
Do you mean it should be reordered to come after `Backend`? Or does "behind" mean a multi level error enum like 
```
enum BackendError {
    Allocation { ... }
    Syscall { ... }
    Other { ... }
}
```
?

Reply:
No, I mean the `anyhow::Error` field of `ModuleError::Backend` should be used to store this error I think. `anyhow::Error` allows storing any type implementing `std::error::Error`, including the `std::io::Error` you have here.

Reply:
Ah, I see. Yeah, `anyhow::Error` can store an `io::Error`, and we could use it to get rid of the `Allocation` and `Syscall` error variants, and use `ModuleError::Backend` for all cases. 

The difference is that the separate variants make it possible for the code catching the error to distinguish between allocation/syscall and other failures, since it can `match` and tell them apart. If all cases are `anyhow::Error`, the caller can't really disambiguate an allocation failure, from a syscall failure, from other `Backend` errors that already exist.

Reply:
For allocation failure having a separate `ModuleError` variant makes sense to me. For syscall failure, I don't think you would ever want to distinguish it from any other `io::Error`, but if that is necessary it would be possible to add a new error type to cranelift-jit wrapping `io::Error` in that case and then store this new type in `anyhow::Error`.